manticore/core/cpu/x86.py

```diff
@@ -5636,99 +5620,102 @@ def VZEROUPPER(cpu):
-            base = cpu.STACK+4 #skip ret address
-            arguments = [ cpu.read_int(base + (i*4), 32) for i in xrange(argcount) ]
-            try:
-                cpu.EAX = function(model, *arguments)
-            except ConcretizeArgument as cae:
-                assert 0 <= cae.argnum < argcount
-                # concretize here
-                mem_addr = base+cae.argnum*4
-                raise ConcretizeMemory(mem_addr, 32, "Concretizing Function Argument", 'MINMAX')
-
-            cpu.EIP = cpu.pop(32)
-        return cdecl_function
-
-    @staticmethod
-    def stdcall(function):
-        '''Standard calling convention
-            Subroutine arguments are passed on the stack. 
-            Callee is responsible for cleaning up the stack.
-            Return values are stored in the EAX register.
-        '''
-        argcount = function.func_code.co_argcount - 1
-        assert argcount >= 0
-        def stdcall_function(model):
-            cpu = model.current
-            # skip saved EIP on stack
-            base = cpu.STACK+4
-            arguments = [ cpu.read_int(base+(pos*4), 32) for pos in xrange(argcount) ]
-            try:
-                cpu.EAX = function(model, *arguments)
-            except ConcretizeArgument as cae:
-                assert 0 <= cae.argnum < argcount
-                # concretize here
-                mem_addr = base+cae.argnum*4
-                raise ConcretizeMemory(mem_addr, 32, "Concretizing Function Argument", 'MINMAX')
-
-            cpu.EIP = cpu.pop(32)
-            cpu.STACK += argcount*4
-        return stdcall_function
-
-    @staticmethod
-    def thiscall(function):
-        pass
+    i386 Linux system call ABI
+    '''
+    def syscall_number(self):
+        return self._cpu.EAX
 
-    @staticmethod
-    def vectorcall(function):
-        pass
+    def get_arguments(self):
+        for reg in ('EBX', 'ECX', 'EDX', 'ESI', 'EDI', 'EBP'):
+            yield reg
+
+    def write_result(self, result):
+        self._cpu.EAX = result
+
+class AMD64LinuxSyscallAbi(SyscallAbi):
+    '''
+    AMD64 Linux system call ABI
+
+    Floating point or wide arguments that deviate from the norm are not yet
+    supported.
+    '''
+    def syscall_number(self):
+        return self._cpu.RAX
+
+    def get_arguments(self):
+        for reg in ('RDI', 'RSI', 'RDX', 'R10', 'R8', 'R9'):
+            yield reg
+
+    def write_result(self, result):
+        self._cpu.RAX = result
+    
+
+class I386CdeclAbi(Abi):
+    '''
+    i386 cdecl function call semantics
+    '''
+    def get_arguments(self):
+        bwidth = self._cpu.address_bit_size / 8
+        offset = self._cpu.ESP
+        while True:
+            offset += bwidth
+            yield offset
+
+    def write_result(self, result):
+        if result is not None:
+            self._cpu.EAX = result
+
+        self._cpu.EIP = self._cpu.pop(self._cpu.address_bit_size)
+        
+class I386StdcallAbi(I386CdeclAbi):
```

Comment:
i see how its useful to inherit like this for get_arguments, but i don't think this inheritance actually makes sense; stdcall and cdecl should be on the same level of the class hierarchy. since arm, x86 and x64 all need some sort of notion of popping args off the stack and there is code duplication for this anyway right now, would we be able to move that up into Abi class itself? remember, cpu.SP is aliased and will work on all cpus.